Index: Classes/Utility/HttpUtility.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- Classes/Utility/HttpUtility.php	(date 1505045171000)
+++ Classes/Utility/HttpUtility.php	(date 1515879118000)
@@ -14,6 +14,8 @@
  * The TYPO3 project - inspiring people to share!
  */

+use Psr\Http\Message\ResponseInterface;
+
 /**
  * HTTP Utility class
  */
@@ -119,4 +121,22 @@
             (isset($urlParts['query']) ? '?' . $urlParts['query'] : '') .
             (isset($urlParts['fragment']) ? '#' . $urlParts['fragment'] : '');
     }
+
+    /**
+     * Send Response to client and exit
+     *
+     * @param ResponseInterface $response
+     * @internal not part of public/stable API yet
+     */
+    public static function sendResponse(ResponseInterface $response)
+    {
+        if (!headers_sent()) {
+            header('HTTP/' . $response->getProtocolVersion() . ' ' . $response->getStatusCode() . ' ' . $response->getReasonPhrase());
+            foreach ($response->getHeaders() as $name => $values) {
+                header($name . ': ' . implode(', ', $values));
+            }
+        }
+        echo $response->getBody()->__toString();
+        exit;
+    }
 }
Index: Configuration/DefaultConfiguration.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- Configuration/DefaultConfiguration.php	(date 1505045171000)
+++ Configuration/DefaultConfiguration.php	(date 1515879118000)
@@ -73,6 +73,7 @@
         'folderCreateMask' => '2775',
         'features' => [
             'security.frontend.keepSessionDataOnLogout' => false,
+            'redirects.hitCount' => false,
         ],
         'createGroup' => '',
         'sitename' => 'TYPO3',
